### PR TITLE
COMMERCE-5629 AccountSelector consolidation and correct styling

### PR DESCRIPTION
**Pull request by gianmarcobrunialti**
How it looked like:
Screenshot 2021-02-25 at 10 23 53

How it'll look like:
Screenshot 2021-02-25 at 10 23 07

### CODE REVIEW

**Comment by gianmarcobrunialti** [comment]
ci:test:sf

**Comment by gianmarcobrunialti** [comment]
ci:test:relevant

File: modules/apps/commerce/commerce-theme-minium/commerce-theme-minium/src/css/minium/minium.scss [hunk 1]
```diff
@@ -5,7 +5,7 @@
 @import './mixins';
 @import './utilities';
 
-@import 'overrides/account_selector';
+// @import 'overrides/account_selector';
```
Comment (wincent) [inline modules/apps/commerce/commerce-theme-minium/commerce-theme-minium/src/css/minium/minium.scss:8]: We should not commit commented-out code.

File: modules/apps/commerce/commerce-frontend-js/src/main/resources/META-INF/resources/components/account_selector/AccountSelector.js [hunk 2]
```diff
@@ -12,104 +12,68 @@
...
+				else {
+					Liferay.fire(CURRENT_ACCOUNT_UPDATED, {id: account.id});
+
+					setCurrentAccount(account);
+					setCurrentView(VIEWS.ORDERS_LIST);
+					setCurrentOrder(null);
+				}
+			})
+			.catch(showErrorNotification);
+	};
```
Comment (wincent) [inline modules/apps/commerce/commerce-frontend-js/src/main/resources/META-INF/resources/components/account_selector/AccountSelector.js:66]: Thanks for remembering the `catch` 🙏

**Review by wincent — COMMENTED** [review]
No blockers from me, only superficial feedback (because I only looked at it superficially).

Be aware that we'll probably activate a lint for all those string attributes that use JSX interpolation unnecessarily; I made a ticket for that: https://github.com/liferay/liferay-frontend-projects/issues/421

File: modules/apps/commerce/commerce-frontend-js/src/main/resources/META-INF/resources/utilities/constants/strings.js [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+	STAR: '*',
+	TAB: '\t',
+	THREE_SPACES: '   ',
+	TILDE: '~',
+	TRIPLE_PERIOD: '...',
+	TRUE: 'true',
+	UNDERLINE: '_',
+	UTC: 'UTC',
+	UTF8: 'UTF-8',
+};
```
[the remark this answers is not in the document]
Comment (wincent) [reply modules/apps/commerce/commerce-frontend-js/src/main/resources/META-INF/resources/utilities/constants/strings.js:103]: Forgot to mention. Almost everything in here is currently unused.

File: modules/apps/commerce/commerce-frontend-js/src/main/resources/META-INF/resources/components/account_selector/AccountSelector.js (hunk 2 above)
Comment (gianmarcobrunialti) [reply modules/apps/commerce/commerce-frontend-js/src/main/resources/META-INF/resources/components/account_selector/AccountSelector.js:66]: @wincent I'd never live without it.

File: modules/apps/commerce/commerce-frontend-js/src/main/resources/META-INF/resources/utilities/constants/strings.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (gianmarcobrunialti) [reply modules/apps/commerce/commerce-frontend-js/src/main/resources/META-INF/resources/utilities/constants/strings.js:103]: @wincent 
*TL;DR* I'll remove this.

Yet if you tell me we have a StringPool in JS, we may want to make use of that if needed.

I'll answer this one only Plain copy-paste of our Java StringPool with some neat-up. I went for the simple idea that we have quite a bit of places in Commerce where we compose strings with, say, white spaces, and other things... and and it may not be memory pressuring, but it's nicer to have a single address in memory to refer to than one and the same string/char re-parsed dozens of times. :D

File: modules/apps/commerce/commerce-frontend-js/src/main/resources/META-INF/resources/components/account_selector/views/OrdersListView.js [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+						}
+
+						return (
+							<OrdersTable
+								orders={items}
+								selectOrderURL={selectOrderURL}
+							/>
+						);
+					}}
+					placeholder={'search-order'}
```
[the remark this answers is not in the document]
Comment (gianmarcobrunialti) [reply modules/apps/commerce/commerce-frontend-js/src/main/resources/META-INF/resources/components/account_selector/views/OrdersListView.js:78]: @wincent 
For this whole family of corrections: the `prop={'...'}` is an automation of the IDE. Personally I prefer it as `props` are not `HTML attributes` and for code look-up/inspections the 2 distinct syntaxes do mark such a... distinction. But yet I'll convert it if it's so "disturbing". :D

Nonetheless, the whole of this component has basically stayed untouched from the version that came to my hands, save some code-splitting to allow for better testability and more fine-grained unit test coverage. It'll certainly undergo a deeper refactor.

**Comment by gianmarcobrunialti** [comment]
ci:test:relevant

**Comment by gianmarcobrunialti** [comment]
ci:test:sf

**Comment by gianmarcobrunialti** [comment]
ci:test:sf

**Comment by gianmarcobrunialti** [comment]
ci:test:relevant

File: modules/apps/commerce/commerce-frontend-js/src/main/resources/META-INF/resources/utilities/constants/strings.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (wincent) [reply modules/apps/commerce/commerce-frontend-js/src/main/resources/META-INF/resources/utilities/constants/strings.js]: > but it's nicer to have a single address in memory to refer to than one and the same string/char re-parsed dozens of times

That sounds very definitely wrong to me.

If you can be bothered, you could benchmark to try and prove a difference, but if you want to save the effort I can tell you that _you won't find any significant difference_. Any difference you find will be buried inside the margins of statistical noise, but note that _even_ if you could measure it, you'd probably get the exact opposite of the result you're expecting; if you "save" the parsing time of re-parsing something like `'/'` everywhere its used, remember that you'll _instead_ be spending time re-parsing constant names like `FORWARD_SLASH`, and worse than that, invoking the module machinery to access the `StringPool` object and then doing a property look up for every access. It's actually doing _more_ work, not less.

That means that _performance_ shouldn't be the dimension you use to add a pattern like this (and if you really did care about it, you wouldn't ship dozens of strings that you aren't even going to reference 😉 ); it should be _readability_ and _writeability_. So, on the readability/writeability axis, things like `BLANK` are very obviously worse (due to ambiguity), things like `FORWARD_SLASH` are going to contribute towards needless wrapping, and things like `INVERTED_QUESTION` are a win because nobody wants to have to remember how to write that. ie. in here you've got the good, the bad, and the ugly.

[the remark this answers is not in the document]
Comment (gianmarcobrunialti) [reply modules/apps/commerce/commerce-frontend-js/src/main/resources/META-INF/resources/utilities/constants/strings.js]: Just to understand though, set aside the non-existent significant execution time differences... I may get that the property look-up in something like an exported (and re-exported)`StringPool` object may not be ideal, especially 'cause 1% of it was used. But still each and all of those keys could actually be exported with a simple `export const CONSTANT = '...';` and at that point I'd see my argument slightly more sound, given tree-shaking and obfuscation (the `CONSTANT` may become a simple `a`) which should take place at build-time (hence making the parser's life easier at exec time -- though insignificantly, I get it).

Correct or...? Just out of discussion.

File: modules/apps/commerce/commerce-frontend-js/src/main/resources/META-INF/resources/components/account_selector/views/OrdersListView.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (wincent) [reply modules/apps/commerce/commerce-frontend-js/src/main/resources/META-INF/resources/components/account_selector/views/OrdersListView.js]: It's not "disturbing". It's just that we (Liferay) strongly value consistency so having mixed style within local regions of a file like you have in this PR is a no-go (Brian may outright reject it if he looks closely enough), and broader consistency across the entire code base is desirable too, regardless of preferences, so liferay-portal is overwhelmingly inclined towards `"x"` over `{'x'}` and I'd recommend you just go with that flow.

I wouldn't worry too much about it though as we will auto-enforce anyway so that people don't have to fret over it.

File: modules/apps/commerce/commerce-frontend-js/src/main/resources/META-INF/resources/utilities/constants/strings.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (wincent) [reply modules/apps/commerce/commerce-frontend-js/src/main/resources/META-INF/resources/utilities/constants/strings.js]: I like the argument from tree-shaking (even though it is still only a theoretical benefit for us at this point because we don't/can't really do it properly due to the way we dynamically bundle resources — ie. due to dynamism of portlets, we can't do ahead-of-time tree-shaking, and we're stuck with module-level granularity for anything we do at runtime in the current system).

But _even if_ your compiler is smart enough to turn `'/'` into an `a`, that `a` is still coming to you via a module import which is not free. You're paying a relatively large amount of overhead just for the pleasure of doing `import {anything} from 'anything-else'`. As an example of the current architecture, look at all the stuff we do just to set-up an AMD module and provide it with access to its dependencies (and _whatever_ we do in the future will have _some_ residual amount of boilerplate-ish overhead just to wire things up):

```js
Liferay.Loader.define("layout-content-page-editor-web@3.0.0/page_editor/core/createDNDBackend", ['module', 'exports', 'require', 'frontend-js-react-web$react-dnd-html5-backend'], function (module, exports, require) {
  var define = undefined;
  var global = window;
  {
    Object.defineProperty(exports, "__esModule", {
      value: true
    });
    exports.default = createDNDBackend;

    // (omitted... the code you actually care about)
  }
});
```

(Just a module selected at random from liferay-portal which imports one thing and exports one thing.)

So, I don't think performance is the right angle to be looking at this from for these tiny strings. It's still readability/writability like I said earlier, and when you look at it through that lens, it's pretty easy to make a case that the indirection of things like `StringPool.BLANK` is harmful, in contrast to counter-examples like `RAQUO_CHAR` and `INVERTED_EXCLAMATION`, which are much easier to defend.

[the remark this answers is not in the document]
Comment (gianmarcobrunialti) [reply modules/apps/commerce/commerce-frontend-js/src/main/resources/META-INF/resources/utilities/constants/strings.js]: Crystal clear and thanks for your time to reply. Yeah I was kind of expecting that the 2 possible counter-arguments would indeed be: "that's Liferay" and (independently from Liferay) "AMD doesn't come for free". Modules are still objects at the end of the day and not having tree-shaking at our disposal (at the moment?), for sure is painful. 

I decidedly didn't go on to defend the readability part because: 1) I really wanted more substantial argumentation related to performance, our infra and the inherent why's. :p 2) I was fully agreeing with you ever since the very beginning, being a bare copy-paste of the StringPool Java class that I wasn't even supposed to commit in full.

So, that's why I thank you twice for taking the time to bear with me on the topic. I owe you a 🍺 .

[the remark this answers is not in the document]
Comment (wincent) [reply modules/apps/commerce/commerce-frontend-js/src/main/resources/META-INF/resources/utilities/constants/strings.js]: > being a bare copy-paste of the StringPool Java class

Well, TIL... this lead me to read the Java implementation today which I'd only ever seen cursorily before, and I was horrified to see how all those names that I criticized are actually from there 😂. Likewise, I looked in the related `CharPool` class, and will file that away for another day if it comes up. 😁

**Comment by ivicac** [comment]
ci:forward
